File: Nov2020/25Nov/OddEvenLinkedList.java
/*
Problem Name: 328. Odd Even Linked List
Problem Link: https://leetcode.com/problems/odd-even-linked-list/
*/
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode oddEvenList(ListNode head) {
        int parity = 0;
        ListNode odd = new ListNode(-1), even = new ListNode(-1), oddTrav = odd, evenTrav = even;
        while(head != null){
            if(parity%2 == 0){
                evenTrav.next = head;
                evenTrav = evenTrav.next;
            }
            else{
                oddTrav.next = head;
                oddTrav = oddTrav.next;
            }
            parity++;
            head = head.next;
        }
        evenTrav.next = odd.next;
        oddTrav.next = null;
        return even.next;
    }
}

File: Nov2020/25Nov/OddEvenLinkedList.cpp
/*
Problem name: 328. Odd Even Linked List
Problem link: https://leetcode.com/problems/odd-even-linked-list/
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        
        if(head==NULL || head->next==NULL|| head->next->next==NULL)  return head;
        
        ListNode *odd,*even,*evenf;
        odd=head;
        even=head->next;
        evenf=head->next;
        
        while(even->next!=NULL){
            odd->next=even->next;
            odd=even->next;
            if(odd->next!=nullptr){
                even->next=odd->next;
                even=odd->next;
            }
            else{
                even->next=nullptr;
            }
            
        }
        odd->next=evenf;
      
        return head;
    }
};
